### ISSUE TITLE
Security vulnerability: user can become a channel operator by accident

### ISSUE THREAD

**Issue by cyrmax** [BearWare/TeamTalk5#2201]
### Description

This problem occurs because the information about channel operators is stored in the form where only the channel ID is recorded.
So if the administrator deletes some channel and creates a new one with the same ID, all users who had some extra rights to the old channel immediately get the same rights in the new one.

Not checked yet, but I think that with other extra rights and per user settings this situation will be the same.

### Application

- [ ] qtTeamTalk
- [ ] TeamTalkAndroid
- [ ] iTeamTalk
- [ ] TeamTalkClassic
- [X] TeamTalkServer

### Platform

- [X] Windows
- [X] macOS
- [ ] Android
- [ ] iOS
- [X] Linux

### Expected behavior

I suggest that when some channel is being deleted while server is running and when server starts / stops and checks the configuration all records that belong to channels that don't exist anymore should be also deleted from the configuration.

### Actual behavior

When you delete a channel and create a new one and it gets the same ID as the old one, users who were old channel operators become the new channel operators too.

### Steps to reproduce problem

1. Create a permanent channel on your server;
2. Set some user as the channel operator;
3. Open Teamtalk config xml file and write down an ID of the newly created channel;
4. Delete the channel;
5. Create a new permanent channel;
6. Notice that its ID is the same as the deleted one;
7. Ask the user who was the previous channel operator to re-login;
8. Find that now he/she is a new channel operator too.

**Comment by beqabeqa473** [comment]
> Not checked yet, but I think that with other extra rights and per user settings this situation will be the same.
> 
Could you elaborate about this?

**Comment by cyrmax** [comment]
@beqabeqa473 Excuse my hasty unchecked conclusions.
Now I have checked and everything besides this operator thing works without any problems.

**Comment by peter1384** [comment]
This only happens with channel operators and it doesn't happen with other user rights because only channel operator user's ids cunflicts with each other.

**Comment by beqabeqa473** [comment]
@bear101 could you please give me a hint where to look? i'd like to fix that if you don't have enough time. I tried looking, but quick look didn't show up anything.

**Comment by bear101** [comment]
This is where it stores channel-IDs: https://github.com/BearWare/TeamTalk5/blob/master/Library/TeamTalkLib/bin/ttsrv/ServerXML.cpp#L1322

It should store channel-path instead.

If you change it then make sure to write support for the modification in ServerXML::UpdateFile: https://github.com/BearWare/TeamTalk5/blob/master/Library/TeamTalkLib/bin/ttsrv/ServerXML.cpp#L1322

**Comment by beqabeqa473** [comment]
Hi.

I don't think this is good idea.

If we are generating a channel, with random id by request, and if such id will be repeated in the future, old chan OP will become also operator of this new channel.

I think, to fix it correctly, on channel remove, all occurences of that chanid should be removed from all available users.

**Comment by peter1384** [comment]
I agree with @beqabeqa473  here, because when a channel is being removed, all related informations and data should be removed from that channel. I've noticed that when a user is a channel operator and that channel gets removed, the user is stil operator of that channel when looking at their user account, Which is bad.

**Comment by amirmahdifard** [comment]
I think that what @bear101 said should be done, because if a user is the operator of the audio box channel, And the audio box channel got removed and another channel with another name has been created, The user should not become operator, but if a channel made again with the name audio box, the user should become operator again in the channel audio box until some admin manually remove the operator of that channel for the user.

**Comment by amirmahdifard** [comment]
@bear101 hi. I hope you will read this message. around 3 or 4 days ago, I made a private channel in my server for someone, and today I logged in in to my server and saw that person is complaning that why did you make other people operator of my channel with out my pormition, I investigated and found out that that user was operator of channel id 12, wich its name was even difrent, but that person didn't wanted their channel so the channel has been deleted, and now completely another person's private channel's id is appeared to be 12, and when I was a sleep even, they realised that they are both operator and they are now telling me why this person is my channel operator with out my right. Please, as soon as possible, store channel path for this instead of channel id, so the user will only become channel operator if new channel's name is the same as the old one, raather than their ids, only if admin manually removes the operator of that user then that user should not become operator, or fix it as how as you know, but please fix it. Server development is very slow because anyone cannot build the server, and @beqabeqa473 is known to promice about fixing a bug but actually never does it, so please fix it your self in 5.18.1. Thanks

**Comment by BlindMaster24** [comment]
@bear101 ?

**Comment by Nardol** [comment]
@BlindMaster24 I really doubt about the real efficiency of your mentions made these last days except in terme of spam people who follow this repository.

**Comment by BlindMaster24** [comment]
I'll work on this in a few days.
@bear101 @beqabeqa473 @cyrmax @amirmahdifard
Based on the discussion, and especially considering Bjørn's opinion as the developer, what do you all think is the best/most stable solution?
We could go with his suggestion to use the channel path, or we could just delete all operator entries for a specific ID when a channel is removed.
I think that just clearing the entries would be faster to implement and wouldn't require users to make any XML changes.
What do you think?

**Comment by bear101** [comment]
When a channel is deleted then you need to sweep through the user accounts and remove the channel from all the user accounts. Only save the new xml-file if changes were actually made to the configuration.

**Comment by BlindMaster24** [comment]
> When a channel is deleted then you need to sweep through the user accounts and remove the channel from all the user accounts. Only save the new xml-file if changes were actually made to the configuration.

That's what I was planning. Thanks for the comment.

**Comment by BlindMaster24** [comment]
@bear101 @beqabeqa473 @cyrmax @Nardol @amirmahdifard  https://github.com/BearWare/TeamTalk5/pull/2700
